File: php/x402-gateway/src/Robots.php
<?php
declare(strict_types=1);

namespace Profullstack\X402Gateway;

final class Robots
{
    /**
     * robots.txt with the crawlers sorted the way the gateway sorts them.
     * $sitemap: null for <siteUrl>/sitemap.xml, '' to omit.
     */
    public static function txt(string $siteUrl, array $disallow = [], array $allow = [], ?string $sitemap = null, string $path = '/crawl', array $refused = [], array $training = Agents::TRAINING, array $retrieval = Agents::RETRIEVAL, array $comments = []): string
    {
        if ($siteUrl === '') {
            throw new \InvalidArgumentException('robots needs siteUrl');
        }
        $base = rtrim($siteUrl, '/');
        $map = $sitemap ?? "$base/sitemap.xml";
        $welcome = function (string $agent) use ($allow, $disallow): string {
            $lines = ["User-agent: $agent", 'Allow: /'];
            foreach ($allow as $p) {
                $lines[] = "Allow: $p";
            }
            foreach ($disallow as $p) {
                $lines[] = "Disallow: $p";
            }
            return implode("\n", $lines);
        };
        $refuse = fn (string $agent): string => "User-agent: $agent\nDisallow: /";
        $charge = fn (string $agent): string => $refuse($agent) . "\nAllow: $path";
        $lines = array_map(fn ($c) => "# $c", $comments);
        if ($comments !== []) {
            $lines[] = '';
        }
        foreach ($refused as $a) {
            $lines[] = $refuse($a) . "\n";
        }
        foreach ($training as $a) {
            $lines[] = $charge($a) . "\n";
        }
        foreach ($retrieval as $a) {
            $lines[] = $welcome($a) . "\n";
        }
        $lines[] = $welcome('*');
        $lines[] = '';
        if ($map !== '') {
            $lines[] = "Sitemap: $map";
            $lines[] = '';
        }
        return implode("\n", $lines);
    }
}

File: go/x402gateway/robots.go
package x402gateway

import "strings"

// RobotsOptions shapes RobotsTxt.
type RobotsOptions struct {
	SiteURL   string
	Disallow  []string
	Allow     []string
	Sitemap   *string // nil: <siteUrl>/sitemap.xml; pointer to "": omit
	Path      string  // default /crawl
	Refused   []string
	Training  []string // nil: TrainingAgents
	Retrieval []string // nil: RetrievalAgents
	Comments  []string
}

// RobotsTxt writes robots.txt with the crawlers sorted the way the gateway sorts them.
func RobotsTxt(o RobotsOptions) string {
	base := strings.TrimRight(o.SiteURL, "/")
	if o.Path == "" {
		o.Path = "/crawl"
	}
	if o.Training == nil {
		o.Training = TrainingAgents
	}
	if o.Retrieval == nil {
		o.Retrieval = RetrievalAgents
	}
	sitemap := base + "/sitemap.xml"
	if o.Sitemap != nil {
		sitemap = *o.Sitemap
	}
	welcome := func(agent string) string {
		lines := []string{"User-agent: " + agent, "Allow: /"}
		for _, p := range o.Allow {
			lines = append(lines, "Allow: "+p)
		}
		for _, p := range o.Disallow {
			lines = append(lines, "Disallow: "+p)
		}
		return strings.Join(lines, "\n")
	}
	refuse := func(agent string) string { return "User-agent: " + agent + "\nDisallow: /" }
	charge := func(agent string) string { return refuse(agent) + "\nAllow: " + o.Path }

	var lines []string
	for _, c := range o.Comments {
		lines = append(lines, "# "+c)
	}
	if len(o.Comments) > 0 {
		lines = append(lines, "")
	}
	for _, a := range o.Refused {
		lines = append(lines, refuse(a)+"\n")
	}
	for _, a := range o.Training {
		lines = append(lines, charge(a)+"\n")
	}
	for _, a := range o.Retrieval {
		lines = append(lines, welcome(a)+"\n")
	}
	lines = append(lines, welcome("*"), "")
	if sitemap != "" {
		lines = append(lines, "Sitemap: "+sitemap, "")
	}
	return strings.Join(lines, "\n")
}
